airflow/models/serialized_dag.py

```diff
@@ -0,0 +1,214 @@
+
+        self.dag_id = dag.dag_id
+        self.fileloc = dag.full_filepath
+        self.fileloc_hash = self.dag_fileloc_hash(self.fileloc)
+        self.data = SerializedDAG.to_dict(dag)
+        self.last_updated = timezone.utcnow()
+
+    @staticmethod
+    def dag_fileloc_hash(full_filepath: str) -> int:
+        """"Hashing file location for indexing.
+
+        :param full_filepath: full filepath of DAG file
+        :return: hashed full_filepath
+        """
+        # hashing is needed because the length of fileloc is 2000 as an Airflow convention,
+        # which is over the limit of indexing. If we can reduce the length of fileloc, then
+        # hashing is not needed.
+        return int.from_bytes(
+            hashlib.sha1(full_filepath.encode('utf-8')).digest()[-2:], byteorder='big', signed=False)
+
+    @classmethod
+    @db.provide_session
+    def write_dag(cls, dag: 'DAG', min_update_interval: Optional[int] = None, session=None):
+        """Serializes a DAG and writes it into database.
+
+        :param dag: a DAG to be written into database
+        :param min_update_interval: minimal interval in seconds to update serialized DAG
+        :param session: ORM Session
+        """
+        log.debug("Writing DAG: %s to the DB", dag)
+        # Checks if (Current Time - Time when the DAG was written to DB) < min_update_interval
+        # If Yes, does nothing
+        # If No or the DAG does not exists, updates / writes Serialized DAG to DB
+        if min_update_interval is not None:
+            if session.query(exists().where(
+                and_(cls.dag_id == dag.dag_id,
+                     (timezone.utcnow() - timedelta(seconds=min_update_interval)) < cls.last_updated))
+            ).scalar():
+                return
+        session.merge(cls(dag))
+        log.debug("DAG: %s written to the DB", dag)
+
+    @classmethod
+    @db.provide_session
+    def read_all_dags(cls, session=None) -> Dict[str, 'SerializedDAG']:
+        """Reads all DAGs in serialized_dag table.
+
+        :param session: ORM Session
+        :returns: a dict of DAGs read from database
+        """
+        serialized_dags = session.query(cls)
+
+        dags = {}
+        for row in serialized_dags:
+            log.debug("Deserializing DAG: %s", row.dag_id)
+            dag = row.dag
+
+            # Sanity check.
+            if dag.dag_id == row.dag_id:
+                dags[row.dag_id] = dag
+            else:
+                log.warning(
+                    "dag_id Mismatch in DB: Row with dag_id '%s' has Serialised DAG "
+                    "with '%s' dag_id", row.dag_id, dag.dag_id)
+        return dags
+
+    @property
+    def dag(self):
+        """The DAG deserialized from the ``data`` column"""
+        from airflow.serialization import SerializedDAG  # noqa # pylint: disable=redefined-outer-name
+
+        if isinstance(self.data, dict):
```

Comment:
Why would `self.data` have two formats here?

Reply:
This probably isn't needed, but might be for cases where SQLAlchemy doesn't handle the JSON deserialize for us.

We should remove this if we can though, yes.

Reply:
Yes, older versions of Mysql don't support json type and sqlalchemy returns string. Best example is our CI images with MySQL

Reply:
Do you have information about which MySQL version does not support this?  Maybe it is worth abandoning the support for some versions to be able to go further in the future e.g. with JSON index etc.

Reply:
>MySQL as of version 5.7 (MariaDB as of the 10.2 series does not)

https://docs.sqlalchemy.org/en/13/core/type_basics.html#sqlalchemy.types.JSON